### PR TITLE
feat: selective disclosure tooling

### PR DESCRIPTION
**Pull request by sinui0**
This PR introduces new tooling for selective disclosure of HTTP and JSON data. It focuses on the builder pattern to provide an ergonomic API for committing and proving the two formats. I decided to carve verification out of scope for this PR.

This is a fairly rough first pass at implementing this tooling, I expect implementation details to change but hopefully this initial API makes sense.

# Changes
- Introduces the `tlsn-formats` crate which also adds the `NotarizedHttpSession` type.
- Encapsulate functionality in `tlsn-prover` into the `tls` module.
- Modifies `SubstringsProofBuilder` to depend on `TranscriptCommitments` instead of `SessionData`
- Adds `http` module to `tlsn-prover` which provides `HttpProver` helps the user construct a `NotarizedHttpSession`

### CODE REVIEW

**Review by th4s — COMMENTED** [review]
Nice work :rocket:

I think it is a good and flexible design that commitment and proof builders mutably borrow `TranscriptCommitmentBuilder` and `SubstringProofBuilder` and so to say use their knowledge over the format to drive them.

Have some questions remarks here and there.

File: tlsn/tlsn-formats/src/http/proof.rs [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+                self.builder,
+                self.commitments,
+                body,
+                Direction::Sent,
+                &mut self.body_built,
+            )
+        })
+    }
+
+    pub fn build(self) -> Result<(), HttpProofBuilderError> {
```
Comment (themighty1) [inline tlsn/tlsn-formats/src/http/proof.rs:197]: need a doc comment here

File: tlsn/tlsn-core/src/transcript.rs [hunk 2]
```diff
@@ -108,6 +108,22 @@ impl RedactedTranscript {
...
     }
+
+    /// Sets all bytes in the transcript which were redacted in the given range.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - The value to set the redacted bytes to
+    /// * `range` - The range of redacted bytes to set
+    pub fn set_redacted_range(&mut self, value: u8, range: Range<usize>) {
+        for range in self
```
Comment (themighty1) [inline tlsn/tlsn-core/src/transcript.rs:119]: we can introduce `set.intersection()` to make this easier to read

**Review by themighty1 — CHANGES_REQUESTED** [review]
lgtm, some minor things need to be fixed

**Review by themighty1 — APPROVED** [review]

File: tlsn/tlsn-formats/src/http/proof.rs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+#[derive(Debug)]
+pub struct HttpResponseProofBuilder<'a, 'b: 'a> {
+    builder: &'a mut SubstringsProofBuilder<'b>,
+    commitments: &'a TranscriptCommitments,
+    response: &'a Response,
+    body: Option<&'a Body>,
+    built: &'a mut bool,
+    body_built: bool,
```
[the remark this answers is not in the document]
Comment (th4s) [reply tlsn/tlsn-formats/src/http/proof.rs:229]: Can we then add a comment?

File: tlsn/tlsn-prover/src/tls/mod.rs [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+    /// Creates an HTTP prover.
+    #[cfg(feature = "formats")]
+    pub fn to_http(self) -> Result<HttpProver<http_state::Closed>, HttpProverError> {
+        HttpProver::new(self)
+    }
+
+    /// Starts notarization of the TLS session.
+    ///
+    /// If the verifier is a Notary, this function will transition the prover to the next state
+    /// where it can generate commitments to the transcript prior to finalization.
```
[the remark this answers is not in the document]
Comment (th4s) [reply tlsn/tlsn-prover/src/tls/mod.rs:235]: I reread it a few times and it makes sense now. I think I still have to get used to the fact that the default case is having a verifier and no notary.

**Review by th4s — APPROVED** [review]
:+1:

**Pull request merged**